ui/pages/asset/components/asset-chart.tsx

```diff
@@ -0,0 +1,307 @@
+import { useI18nContext } from '../../../hooks/useI18nContext';
+import fetchWithCache from '../../../../shared/lib/fetch-with-cache';
+import { MINUTE } from '../../../../shared/constants/time';
+import AssetPrice from './asset-price';
+import ChartTooltip from './chart-tooltip';
+
+/** Time range units supported by the price API */
+type TimeRange = `${number}D` | `${number}M` | `${number}Y`;
+
+/** A Chart.js plugin that draws a vertical crosshair on hover */
+const crosshairPlugin = {
+  id: 'crosshair',
+  afterEvent(chart: Chart, { event }: { event: ChartEvent }) {
+    chart.crosshairX = event.type === 'mouseout' ? undefined : event.x;
+    chart.draw();
+  },
+  afterDraw(chart: Chart) {
+    if (chart.crosshairX !== undefined) {
+      const { y } = chart.scales;
+      chart.ctx.beginPath();
+      chart.ctx.moveTo(chart.crosshairX, y.getPixelForValue(y.max));
+      chart.ctx.lineWidth = 2;
+      chart.ctx.strokeStyle = '#BBC0C5';
+      chart.ctx.lineTo(chart.crosshairX, y.getPixelForValue(y.min));
+      chart.ctx.stroke();
+    }
+  },
+};
+
+Chart.register(
+  LinearScale,
+  PointElement,
+  LineElement,
+  Filler,
+  Decimation,
+  crosshairPlugin,
+);
+
+const initialChartOptions: ChartOptions<'line'> = {
+  normalized: true,
+  parsing: false,
+  aspectRatio: 2.3,
+  layout: { autoPadding: false, padding: 0 },
+  animation: { duration: 0 },
+  fill: true,
+  borderColor: '#0376c9',
+  backgroundColor: ({ chart }: { chart: Chart }) => {
+    const gradient = chart.ctx.createLinearGradient(0, 0, 0, chart.height);
+    gradient.addColorStop(0, '#0376c94C');
+    gradient.addColorStop(1, '#D9D9D900');
+    return gradient;
+  },
+  elements: {
+    line: { borderWidth: 2 },
+    point: { pointStyle: false },
+  },
+  plugins: {
+    decimation: {
+      algorithm: 'lttb',
+      samples: 150,
+      threshold: 150,
+      enabled: true,
+    },
+  },
+};
+
+// A chart showing historic prices for a native or token asset
+const AssetChart = ({
+  address,
+  currentPrice,
+}: {
+  address: string;
+  currentPrice: number;
+}) => {
+  const t = useI18nContext();
+  const chainId = hexToDecimal(useSelector(getCurrentChainId));
+  const currency = useSelector(getCurrentCurrency);
+
+  const [data, setData] = useState<ChartData<'line', Point[]>>();
+  const [loading, setLoading] = useState<boolean>(true);
+  const [timeRange, setTimeRange] = useState<TimeRange>('1D');
+  const [chartOptions, setChartOptions] = useState(initialChartOptions);
+
+  const chartRef = useRef<Chart<'line', Point[]>>();
+  const priceRef = useRef<{
+    setPrices: (_: {
+      price?: number;
+      comparePrice?: number;
+      date?: number;
+    }) => void;
+  }>();
+  const maxPriceTooltip = useRef<{
+    setTooltip: (_: { xAxisPercent: number; price: number }) => void;
+  }>();
+  const minPriceTooltip = useRef<{
+    setTooltip: (_: { xAxisPercent: number; price: number }) => void;
+  }>();
```

Comment:
These refs are for performance, I think.  I want to re-render as little as possible on hover/resize.  Instead of storing this state on this parent, I've pushed this state into the (`AssetPrice`, `ChartTooltip`) children, the only components that change on (hover, resize) respectively.  But since the hover/resize events occur on their sibling chart component, you need a ref to push it from parent -> child.

Let me know if this optimization doesn't make sense.